aws/resource_aws_elasticsearch_domain_policy.go

```diff
@@ -93,19 +93,27 @@ func resourceAwsElasticSearchDomainPolicyUpsert(d *schema.ResourceData, meta int
 
 func resourceAwsElasticSearchDomainPolicyDelete(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).esconn
+	domainName := d.Get("domain_name").(string)
 
 	_, err := conn.UpdateElasticsearchDomainConfig(&elasticsearch.UpdateElasticsearchDomainConfigInput{
-		DomainName:     aws.String(d.Get("domain_name").(string)),
+		DomainName:     aws.String(domainName),
 		AccessPolicies: aws.String(""),
 	})
 	if err != nil {
+		log.Printf("[WARN] ElasticSearch Domain Policy for %s delete failed: %s", domainName, err)
+
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" {
+			log.Printf("[WARN] ElasticSearch Domain %q not found, removing", domainName)
+			d.SetId("")
+			return nil
+		}
```

Comment:
The condition here in `Delete()` isn't necessary, because by default `apply` always refreshes the state and runs `Read()`, so the above condition should be sufficient and more aligned with behaviour of majority of existing resources in the codebase.